core/jobs/batch_jobs/story_node_jobs.py

```diff
@@ -0,0 +1,183 @@
+# coding: utf-8
+#
+# Copyright 2023 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs used for populating story node."""
+
+from __future__ import annotations
+
+import logging
+
+from core import utils
+from core.jobs import base_jobs
+from core.jobs.io import ndb_io
+from core.jobs.transforms import job_result_transforms
+from core.jobs.types import job_run_result
+from core.platform import models
+
+import apache_beam as beam
+import result
+from typing import Tuple
+
+MYPY = False
+if MYPY: # pragma: no cover
+    from mypy_imports import datastore_services
+    from mypy_imports import story_models
+    from mypy_imports import topic_models
+
+(story_models, topic_models) = models.Registry.import_models([
+    models.Names.STORY, models.Names.TOPIC])
+datastore_services = models.Registry.import_datastore_services()
+
+
+class PopulateStoryNodeJob(base_jobs.JobBase):
+    """Job that migrates story nodes, for serial chapter launch feature"""
+
+    DATASTORE_UPDATES_ALLOWED = True
+
+    def _update_story_node(
+            self, story_model: story_models.StoryModel
+    ) -> result.Result[
+        story_models.StoryModel,
+        Tuple[str, Exception]
+    ]:
+        """Populate the 5 new fields in each story node of the StoryModel
+        instance, namely: status, planned_publication_date_msecs,
+        last_modified_msecs, first_publication_date_msecs, unpublishing_reason.
+
+        Args:
+            story_model: StoryModel. The story whose nodes have to be
+                populated.
+
+        Returns:
+            Result(StoryModel, (str, Exception)). Result containing the
+            updated StoryModel instance to be put.
+        """
+
+        nodes = story_model.story_contents['nodes']
+
+        try:
+            with datastore_services.get_ndb_context():
+                topic_model = topic_models.TopicModel.get(
```

Comment:
Why do we need to make get request here? I'm assuming the total number of request would be O(m*n) here [m=nuber of stories in topic, n=number of topics]

Can we make parallel request in pipe to fetch all the topics at once and create a tuple of (topic, [stories]) and traverse through each item of the tuple?

Reply:
Yes that can be done. Actually I am writing beam jobs for the first time, so I wrote this in a simplistic manner. I am not sure of how we can create a tuple of topic and stories array using the PCollections for topic models and story models. It will be helpful if you can share some clues about it.

Reply:
@Kej-r03 Please check other *_jobs.py file for reference (e.g. story_migration_jobs.py). Also, go through the wiki page you'll find important information. Writing the first job feels daunting but it's not that difficult, I think it's just that the syntax/pattern is a little different than general Python code which makes it look difficult (not sure!). Let me know if you have any specific questions, I'll be more than happy to help! :)

Reply:
Also, you can read the official ref doc which contains good examples:
https://beam.apache.org/documentation/transforms/python/
https://beam.apache.org/documentation/transforms/python/elementwise/

Reply:
Done